File: android/app/src/main/java/in/anairapos/app/AnairaLocalDbPlugin.java
package in.anairapos.app;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.PluginMethod;

import org.json.JSONObject;

@CapacitorPlugin(name = "AnairaLocalDb")
public class AnairaLocalDbPlugin extends Plugin {
    private LocalDb db;

    private LocalDb db() {
        if (db == null) db = new LocalDb(getContext());
        return db;
    }

    @PluginMethod
    public void open(PluginCall call) {
        db().getWritableDatabase();
        JSObject out = new JSObject();
        out.put("success", true);
        out.put("driver", "android-sqlite");
        out.put("restaurantId", call.getString("restaurantId", ""));
        call.resolve(out);
    }

    @PluginMethod
    public void put(PluginCall call) {
        String restaurantId = call.getString("restaurantId", "");
        String entity = call.getString("entity", "");
        String id = call.getString("id", "");
        JSObject data = call.getObject("data", new JSObject());
        if (restaurantId.isEmpty() || entity.isEmpty() || id.isEmpty()) {
            call.reject("restaurantId, entity and id are required"); return;
        }
        ContentValues values = new ContentValues();
        values.put("key", restaurantId + ":" + entity + ":" + id);
        values.put("restaurant_id", restaurantId);
        values.put("entity", entity);
        values.put("record_id", id);
        values.put("data", data.toString());
        db().getWritableDatabase().insertWithOnConflict("records", null, values, SQLiteDatabase.CONFLICT_REPLACE);
        JSObject out = new JSObject(); out.put("success", true); call.resolve(out);
    }


    @PluginMethod
    public void list(PluginCall call) {
        String restaurantId = call.getString("restaurantId", "");
        String entity = call.getString("entity", "");
        Cursor c = db().getReadableDatabase().query("records", new String[]{"data"}, "restaurant_id=? AND entity=?", new String[]{restaurantId, entity}, null, null, "updated_at ASC");
        try {
            com.getcapacitor.JSArray out = new com.getcapacitor.JSArray();
            while (c.moveToNext()) out.put(new JSONObject(c.getString(0)));
            JSObject result = new JSObject();
            result.put("records", out);
            call.resolve(result);
        } catch (Exception e) { call.reject(e.getMessage()); }
        finally { c.close(); }
    }

    @PluginMethod
    public void remove(PluginCall call) {
        String restaurantId = call.getString("restaurantId", "");
        String entity = call.getString("entity", "");
        String id = call.getString("id", "");
        db().getWritableDatabase().delete("records", "key=?", new String[]{restaurantId + ":" + entity + ":" + id});
        JSObject result = new JSObject(); result.put("success", true); call.resolve(result);
    }

    @PluginMethod
    public void get(PluginCall call) {
        String restaurantId = call.getString("restaurantId", "");
        String entity = call.getString("entity", "");
        String id = call.getString("id", "");
        Cursor c = db().getReadableDatabase().query("records", new String[]{"data"}, "key=?", new String[]{restaurantId + ":" + entity + ":" + id}, null, null, null);
        try {
            if (!c.moveToFirst()) { call.resolve(new JSObject()); return; }
            call.resolve(JSObject.fromJSONObject(new JSONObject(c.getString(0))));
        } catch (Exception e) { call.reject(e.getMessage()); }
        finally { c.close(); }
    }

    static class LocalDb extends SQLiteOpenHelper {
        LocalDb(android.content.Context context) { super(context, "anaira_local.db", null, 1); }
        @Override public void onCreate(SQLiteDatabase db) {
            db.execSQL("CREATE TABLE records (key TEXT PRIMARY KEY, restaurant_id TEXT NOT NULL, entity TEXT NOT NULL, record_id TEXT NOT NULL, data TEXT NOT NULL, updated_at INTEGER NOT NULL DEFAULT (strftime('%s','now')))");
            db.execSQL("CREATE INDEX idx_records_restaurant_entity ON records(restaurant_id, entity)");
        }
        @Override public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) { }
    }
}

File: android-native/AnairaLocalDbPlugin.kt
package in.anairapos.app

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject

@CapacitorPlugin(name = "AnairaLocalDb")
class AnairaLocalDbPlugin : Plugin() {
  private lateinit var helper: LocalDbHelper

  override fun load() {
    super.load()
    helper = LocalDbHelper(context)
  }

  @com.getcapacitor.PluginMethod
  fun open(call: PluginCall) {
    helper.writableDatabase
    val ret = JSObject()
    ret.put("success", true)
    ret.put("driver", "android-sqlite")
    ret.put("version", LocalDbHelper.VERSION)
    call.resolve(ret)
  }

  @com.getcapacitor.PluginMethod
  fun put(call: PluginCall) {
    val restaurantId = call.getString("restaurantId") ?: return call.reject("restaurantId is required")
    val entity = call.getString("entity") ?: return call.reject("entity is required")
    val id = call.getString("id") ?: return call.reject("id is required")
    val data = call.getObject("data") ?: JSObject()
    helper.put(restaurantId, entity, id, data)
    call.resolve()
  }

  @com.getcapacitor.PluginMethod
  fun get(call: PluginCall) {
    val restaurantId = call.getString("restaurantId") ?: return call.reject("restaurantId is required")
    val entity = call.getString("entity") ?: return call.reject("entity is required")
    val id = call.getString("id") ?: return call.reject("id is required")
    call.resolve(helper.get(restaurantId, entity, id) ?: JSObject())
  }

  @com.getcapacitor.PluginMethod
  fun list(call: PluginCall) {
    val restaurantId = call.getString("restaurantId") ?: return call.reject("restaurantId is required")
    val entity = call.getString("entity") ?: return call.reject("entity is required")
    val limit = call.getInt("limit", 500) ?: 500
    val ret = JSObject()
    ret.put("records", JSArray(helper.list(restaurantId, entity, limit)))
    call.resolve(ret)
  }

  @com.getcapacitor.PluginMethod
  fun remove(call: PluginCall) {
    val restaurantId = call.getString("restaurantId") ?: return call.reject("restaurantId is required")
    val entity = call.getString("entity") ?: return call.reject("entity is required")
    val id = call.getString("id") ?: return call.reject("id is required")
    helper.remove(restaurantId, entity, id)
    call.resolve()
  }

  private class LocalDbHelper(ctx: android.content.Context) : SQLiteOpenHelper(ctx, DB_NAME, null, VERSION) {
    override fun onCreate(db: SQLiteDatabase) {
      db.execSQL("CREATE TABLE IF NOT EXISTS records(key TEXT PRIMARY KEY, restaurant_id TEXT NOT NULL, entity TEXT NOT NULL, entity_id TEXT NOT NULL, data TEXT NOT NULL, updated_at TEXT NOT NULL)")
      db.execSQL("CREATE INDEX IF NOT EXISTS idx_records_restaurant_entity ON records(restaurant_id,entity)")
    }
    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) = onCreate(db)
    fun put(restaurantId: String, entity: String, id: String, data: JSObject) {
      val values = ContentValues().apply {
        put("key", "$restaurantId:$entity:$id")
        put("restaurant_id", restaurantId)
        put("entity", entity)
        put("entity_id", id)
        put("data", data.toString())
        put("updated_at", java.time.Instant.now().toString())
      }
      writableDatabase.insertWithOnConflict("records", null, values, SQLiteDatabase.CONFLICT_REPLACE)
    }
    fun get(restaurantId: String, entity: String, id: String): JSObject? {
      readableDatabase.rawQuery("SELECT data FROM records WHERE key=?", arrayOf("$restaurantId:$entity:$id")).use {
        if (!it.moveToFirst()) return null
        return JSObject(it.getString(0))
      }
    }
    fun list(restaurantId: String, entity: String, limit: Int): List<JSONObject> {
      val out = mutableListOf<JSONObject>()
      readableDatabase.rawQuery("SELECT data FROM records WHERE restaurant_id=? AND entity=? ORDER BY updated_at DESC LIMIT ?", arrayOf(restaurantId, entity, limit.toString())).use {
        while (it.moveToNext()) out.add(JSONObject(it.getString(0)))
      }
      return out
    }
    fun remove(restaurantId: String, entity: String, id: String) {
      writableDatabase.delete("records", "key=?", arrayOf("$restaurantId:$entity:$id"))
    }
    companion object { const val DB_NAME = "anaira-pos.db"; const val VERSION = 11 }
  }
}
